### ISSUE TITLE
Type narrowing for relationships in between variables.

### ISSUE THREAD

**Issue by GugelRobin** [microsoft/pyright#10310]
Dear pyright team,

I suspect that this was probably brought up before, but I didn't find any thread.

**Is your feature request related to a problem? Please describe.**
I've had that problem quite a few times already. I have two variables which may be None, and I want to do something while at least one variable is not None, and then check what to do.

Consider the following dummy code:
```python

def val() -> int | None:
    ...

def work(val: int):
    ... 
a = val()
b = val()

while a is not None or b is not None:
    if a is None:
        work(b) # False positive: Type "int | None" is not assignable to type "int"
                #                   "None" is not assignable to "int"
    elif b is None:
        work(a)
    else:
        work(a)
        work(b)
    a = val()
    b = val()
```

**Describe the solution you’d like**
I'd like that pyright understands that if 'a' is None, 'b' cannot be (and vice versa).
Similarily like it works for 'match-case' statements:
```python
match (a, b):
    case (None, None):
        ...
    case (_, None):
        work(a) # a is inferred to be an int
    case (None, _):
        work(b) # b is inferred to be an int
    case _:
        work(a)
        work(b)
```
The following sematically equivalent if-cascade should work just like 'match-case':
```python
if (a is None) and (b is None):
    ...
elif b is None:
    work(a) # Argument of type "int | None" cannot be assigned to parameter "val" of type "int"
elif a is None:
    work(b)
else:
    work(a)
    work(b)
```

I haven't found any proper workaround yet (especially when working with loops), besides littering my code with `assert a is not None` and/or `#type ignore`

Thank you for your effort.

**Comment by erictraut** [comment]
Yes, this has been brought up before. Type narrowing is applied separately to expressions. What you're suggesting here would involve tracking conditional types such as "the type of `a` is `None` if the type the type of `b` is `None` and has not been subsequently reassigned, otherwise the type of `a` is `int`". This sort of conditional type tracking is theoretically possible, but in practice it's computationally infeasible. I don't know of any compiler or type checker in any language that does this. Keep in mind that static type checkers, unlike runtime engines, need to consider all possible code paths and types that can result along every one of those code paths. This results in large combinatorics. Adding conditional types on top of that would be combinatorics on top of combinatorics.

For those reasons, what you're suggesting here is unlikely to ever materialize in pyright or any other static type checker.

There's almost always a way to restructure your code to avoid a reliance on this sort of "entanglement" between two variables. If you don't want to restructure, you can typically aid the type checker by inserting `assert` statements that document your assumptions.

```python
    if a is None:
        assert b is not None
```
